### PR TITLE
Improve vacuumdb performance

### PR DESCRIPTION
**Pull request by bmdoil**
Beginning in PG12 https://github.com/postgres/postgres/commit/e0c2933a7, vacuumdb issues a separate VACUUM command for each targeted table. This change resulted in a performance regression in upstream due to spending O(N^2) time in vac_update_datfrozenxid(). 

PG16 https://github.com/postgres/postgres/commit/a46a701 introduced options to control whether VACUUM runs vac_update_datfrozenxid. This PR backports the upstream commit.

Issuing separate VACUUM's for each table has an even greater performance regression in GPDB due to the dispatch overhead. Instead we refactor how vacuumdb builds VACUUM commands.

When vacuumdb is run without the --jobs/--tables flags, execute a single database wide VACUUM. This restores the behavior of 6X vacuumdb.

When vacuumdb is run with --jobs/--tables, divide the table list evenly among available connections and issue one VACUUM per connection.

Performance on a DB with 30k tables:

main branch:
```
vacuumdb postgres  0.28s user 0.20s system 0% cpu 8:29.80 total
vacuumdb postgres --jobs 2  0.36s user 0.30s system 0% cpu 6:35.34 total
vacuumdb postgres --jobs 8  0.45s user 0.57s system 0% cpu 2:58.57 total
```
PR:
```
vacuumdb postgres  0.01s user 0.00s system 0% cpu 3:39.39 total
vacuumdb postgres --jobs 2  0.00s user 0.01s system 0% cpu 2:24.89 total
vacuumdb postgres --jobs 8  0.01s user 0.00s system 0% cpu 1:37.36 total
```

### CODE REVIEW

**Comment by ashwinstar** [comment]
- Understand GPDB has more impact without and hence also more gain with this PR, though impact must be seen on PostgreSQL as well?
- Would be nice to have more explanation on the batch-size functionality. Also, would love to hear from you on its relevance to upstream as well?

File: src/bin/scripts/vacuumdb.c [hunk 1]
```diff
@@ -118,6 +118,7 @@ main(int argc, char *argv[])
 		{"skip-locked", no_argument, NULL, 5},
 		{"min-xid-age", required_argument, NULL, 6},
 		{"min-mxid-age", required_argument, NULL, 7},
+		{"batch-size", required_argument, NULL, 8},
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:121]: Shouldn't this be `optional_argument`? Pretty funny how even --jobs is a "required_argument". Meh.

File: src/bin/scripts/vacuumdb.c [hunk 2]
```diff
@@ -666,55 +677,97 @@ vacuum_one_database(const ConnParams *cparams,
...
+	 */
+	if (parallel || tables_listed)
 	{
-		const char *tabname = cell->val;
-		ParallelSlot *free_slot;
+		int processedTables = 0;
 
-		if (CancelRequested)
+		cell = dbtables.head;
+		do
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:690]: Hmm this indented diff can be avoided by the same mechanism we use in other parts of the code. For instance, see:
https://github.com/greenplum-db/gpdb/blob/3e9c8d09a64d1a3be8c99f0b64f6b4138bd16847/src/backend/access/brin/brin.c#L529-L531

Comment (soumyadeep2007) [reply src/bin/scripts/vacuumdb.c:690]: Ping me once you fix this ^, it will make it considerably easier to review afterwards.

File: src/bin/scripts/vacuumdb.c (hunk 1 above)
Comment (ashwinstar) [reply src/bin/scripts/vacuumdb.c:121]: I think `required_argument` means this flag needs argument to be supplied (like --jobs <requires argument for number). It's not about if this flag is required or not.

File: src/bin/scripts/vacuumdb.c (hunk 2 above)
Comment (bmdoil) [reply src/bin/scripts/vacuumdb.c]: @soumyadeep2007 It should be much easier to review now.

File: src/bin/scripts/vacuumdb.c [hunk 3]
```diff
@@ -752,6 +752,19 @@ vacuum_one_database(const ConnParams *cparams,
...
+	{
+	/*
+	 * GPDB: Prepare and execute a database wide vacuum in non-parallel mode.
+	 * We use the only connection slot we have, which we know to be free.
+	 * In this mode, the program terminates in case of an error.
+	 */
+		prepare_vacuum_command(&sql, PQserverVersion(slots->connection),
+								vacopts, NULL);
+		run_vacuum_command(slots->connection, sql.data,
+								echo, NULL, progname, parallel);
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:766]: We can move this to the early return.

File: src/bin/scripts/vacuumdb.c [hunk 4]
```diff
@@ -752,6 +752,19 @@ vacuum_one_database(const ConnParams *cparams,
...
 
 	} while (cell != NULL);
+	}
+	else
+	{
+	/*
+	 * GPDB: Prepare and execute a database wide vacuum in non-parallel mode.
+	 * We use the only connection slot we have, which we know to be free.
+	 * In this mode, the program terminates in case of an error.
+	 */
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:762]: Weird indentation.

Let's also add the reasoning: to save on dispatch overhead.

**Review by soumyadeep2007 — COMMENTED** [review]
We need a vacuumdb TAP test for the batch option. We could probably leverage `issues_sql_like`.

File: src/bin/scripts/vacuumdb.c [hunk 5]
```diff
@@ -679,41 +700,57 @@ vacuum_one_database(const ConnParams *cparams,
...
+		 * GPDB: Prepare the vacuum command. If it's not a new command, append
+		 * the table to the query string.
+		 */
+		if (newCmd)
+			prepare_vacuum_command(&sql, PQserverVersion(free_slot->connection),
+									vacopts, tabname);
 		else
-			free_slot = slots;
+			prepare_vacuum_command(&sql, PQserverVersion(free_slot->connection),
+									NULL, tabname);
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/scripts/vacuumdb.c:736]: On main `prepare_vacuum_command` always semicolon-terminates the string buffer `appendPQExpBuffer(sql, " %s;", table);` so that will need to change.

The server handles the commands properly without the explicit semicolon in my tests, but to be defensive I think it makes more sense to append a semicolon in `run_vacuum_command` before we send the query.

**Comment by soumyadeep2007** [comment]
For the upper limit for the batch size, there is no real upper limit on the number of relations that can be specified in a VACUUM command. Thinking more about it, it is a bad experience for users to come and run the VACUUM with a high batch size, only for it to error out with the VMEM error that you were describing to me today.

We can probably cap it at an arbitrary limit, based on the default VMEM settings perhaps.

**Comment by bmdoil** [comment]
> For the upper limit for the batch size, there is no real upper limit on the number of relations that can be specified in a VACUUM command. Thinking more about it, it is a bad experience for users to come and run the VACUUM with a high batch size, only for it to error out with the VMEM error that you were describing to me today.
> 
> We can probably cap it at an arbitrary limit, based on the default VMEM settings perhaps.

I looked into the VMEM issue more. I had `optimizer=on` and `optimizer_analyze_root_partition=on` and it fell over in `leaf_parts_analyzed` on the coordinator. This is a known issue that was addressed in https://github.com/greenplum-db/gpdb/pull/10499 but it seems I've hit an upper limit. 

```
(gdb) bt
#0  0x00005592e69023a4 in _bt_first (scan=0x5592ea1fb0e0, dir=ForwardScanDirection) at nbtsearch.c:922
#1  0x00005592e68ff51f in btgettuple (scan=0x5592ea1fb0e0, dir=ForwardScanDirection) at nbtree.c:327
#2  0x00005592e68f57db in index_getnext_tid (scan=0x5592ea1fb0e0, direction=ForwardScanDirection) at indexam.c:526
#3  0x00005592e68f5962 in index_getnext_slot (scan=0x5592ea1fb0e0, direction=ForwardScanDirection, slot=0x5592ea1f8cc0) at indexam.c:618
#4  0x00005592e68f4623 in systable_getnext (sysscan=0x5592eb091100) at genam.c:450
#5  0x00005592e6b755ee in fetch_leaf_att_stats (leafRelid=39861, leafAttNum=1) at analyzeutils.c:1138
#6  0x00005592e6b75959 in leaf_parts_analyzed (attrelid=26103, relid_exclude=26487, va_cols=0x5592f9651860, elevel=13) at analyzeutils.c:1285
#7  0x00005592e6b6c096 in expand_vacuum_rel (vrel=0x5592ea25a4c0, options=1) at vacuum.c:1045
#8  0x00005592e6b6b02d in vacuum (relations=0x5592e9ec3488, params=0x7fff76ee9510, bstrategy=0x5592ea1f4d60, isTopLevel=true) at vacuum.c:385
#9  0x00005592e6b6ad8d in ExecVacuum (pstate=0x5592ea1192c0, vacstmt=0x5592ea4a5f50, isTopLevel=true, auto_stats=false) at vacuum.c:252
```
**I'm a little surprised that this is being called since I didn't specify `vacuumdb --analyze`.** 

Per the comment in `expand_vacuum_rel` 
```
/*
* GPDB: If you explicitly ANALYZE a partition, also update the
* parent's stats after the partition has been ANALYZEd. (Thanks to
* the code to merge leaf statistics, it should be fast.)
*
```
We don't skip partition children and if optimizer_analyze_root_partition=on, we're descending into `leaf_parts_analyzed`. Is this a bug?
```
		if (optimizer_analyze_root_partition && !skip_this)
		{
                                .......
				/*
				 * Only ANALYZE the parent if the stats can be updated by merging
				 * child stats.
				 */
				if (!leaf_parts_analyzed(parent_relid, child_relid, vrel->va_cols, elevel))
					break;
```
VACUUM without a relation list is fine because we don't call `expand_vacuum_rel` 

After disabling `optimizer_analyze_root_partition` I had no issues running a VACUUM with a list of 30,000 tables.

With this context, `batch-size` feels like a workaround for a bug. I don't think we need to add the option at all.

**Comment by ashwinstar** [comment]
> We don't skip partition children and if optimizer_analyze_root_partition=on, we're descending into leaf_parts_analyzed. Is this a bug?
Maybe this discussion is relevant https://github.com/greenplum-db/gpdb/pull/16218 for this @chrishajas

**Comment by soumyadeep2007** [comment]
> We don't skip partition children and if optimizer_analyze_root_partition=on, we're descending into leaf_parts_analyzed. Is this a bug?

Wow this is a terrible bug. There is no reason for us to do this check `leaf_parts_analyzed` or add to the vacuum list if `analyze` has not even been specified.

We should guard like this: `if ((options & VACOPT_ANALYZE) && optimizer_analyze_root_partition && !skip_this)` and then that should solve this problem.

> After disabling optimizer_analyze_root_partition I had no issues running a VACUUM with a list of 30,000 tables.

> With this context, batch-size feels like a workaround for a bug. I don't think we need to add the option at all.

I agree.

**Review by soumyadeep2007 — COMMENTED** [review]
We need TAP tests to ensure:
1. A single db wide VACUUM is fired when --jobs/tables is unspecified,
2. We have a db with tables - see if we batch the tables according to --jobs, if only --jobs is specified.
3. We have a db with tables and the --tables list is specified - see if we batch the tables according to --jobs

File: src/bin/scripts/vacuumdb.c [hunk 6]
```diff
@@ -686,41 +708,56 @@ vacuum_one_database(const ConnParams *cparams,
...
+		*
+		* If not in parallel mode, this terminates the
+		* program in case of an error.  (The parallel case handles query
+		* errors in ProcessQueryResult through GetIdleSlot.)
+		*/
+		if (++processedTables == listSize)
+		{
+			run_vacuum_command(free_slot->connection, &sql,
+									echo, tabname, progname, parallel);
+			processedTables = 0;
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/scripts/vacuumdb.c:756]: `prepare_vacuum_command` resets the buffer
```
prepare_vacuum_command(PQExpBuffer sql, int serverVersion,
					   vacuumingOptions *vacopts, const char *table)
{
	const char *paren = " (";
	const char *comma = ", ";
	const char *sep = paren;

	resetPQExpBuffer(sql);
```

File: src/bin/scripts/vacuumdb.c [hunk 7]
```diff
@@ -932,31 +969,36 @@ prepare_vacuum_command(PQExpBuffer sql, int serverVersion,
...
 static void
-run_vacuum_command(PGconn *conn, const char *sql, bool echo,
+run_vacuum_command(PGconn *conn, PQExpBuffer sql, bool echo,
 				   const char *table, const char *progname, bool async)
 {
 	bool		status;
 
+	/* Ensure query string is semicolon-terminated */
+	if (sql->data[sql->len - 1] != ';')
+		appendPQExpBufferChar(sql, ';');
```
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/scripts/vacuumdb.c:991]: > Or you can make it more formal by wrapping `appendPQExpBufferChar(sql, ';');` in a function `finalize_vacuum_command` and adding a comment to `prepare_vacuum_command`: `GPDB: We don't semi-colon-terminate. finalize_vacuum_command() must be called for that to happen.`

This makes sense. We could formalize the append also, which means `prepare_vacuum_command` is only tasked with building the initial `VACUUM` options.

```
/*
 * GPDB: append table to the query string
 */
static void
append_vacuum_command(PQExpBuffer sql, char* table)
{
	appendPQExpBuffer(sql, ", %s", table);
}

/*
 * GPDB: semi-colon terminate the query string
 */
static void
finalize_vacuum_command(PQExpBuffer sql)
{
	appendPQExpBufferChar(sql, ';');
}
```

[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/bin/scripts/vacuumdb.c:991]: +1 to formalizing append.

File: src/bin/scripts/vacuumdb.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/bin/scripts/vacuumdb.c:756]: Ah okay got it.

File: src/bin/scripts/vacuumdb.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (bmdoil) [reply src/bin/scripts/vacuumdb.c]: Done!

File: src/bin/scripts/vacuumdb.c [hunk 8]
```diff
@@ -686,39 +711,54 @@ vacuum_one_database(const ConnParams *cparams,
...
+		if (++processedTables == listSize)
+		{
+			finalize_vacuum_command(&sql);
+			run_vacuum_command(free_slot->connection, &sql,
+									echo, tabname, progname, parallel);
+			processedTables = 0;
+		}
 
 		cell = cell->next;
 	} while (cell != NULL);
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:764]: After this loop ends you need to process the "last batch" in the case that you had ntables not divisible by jobs.

For instance consider `ntables = ntups = 10` and `jobs = concurrentCons = 4`, then the last batch of t9 and t10 would still be "prepared", but not "run"

So outside the loop, you need to do something like:
```
  if (processedTables < listSize)
  {
       finalize_vacuum_command(&sql);
	   run_vacuum_command(free_slot->connection, &sql,
									echo, tabname, progname, parallel);
   }
```

We should also add a test case covering this.

File: src/bin/scripts/vacuumdb.c [hunk 9]
```diff
@@ -931,19 +975,35 @@ prepare_vacuum_command(PQExpBuffer sql, int serverVersion,
 				appendPQExpBufferStr(sql, " ANALYZE");
 		}
 	}
+}
+
+/*
+ * GPDB: append table to the query string
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:981]: s/to the query string/to the vacuum command

File: src/bin/scripts/vacuumdb.c [hunk 10]
```diff
@@ -931,19 +975,35 @@ prepare_vacuum_command(PQExpBuffer sql, int serverVersion,
...
+ */
+static void
+append_vacuum_command(PQExpBuffer sql, char* table)
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:984]: s/append_vacuum_command/append_to_vacuum_command

File: src/bin/scripts/vacuumdb.c [hunk 11]
```diff
@@ -931,19 +975,35 @@ prepare_vacuum_command(PQExpBuffer sql, int serverVersion,
...
- * sending the command; else, wait for it to finish.
+ * Send a vacuum/analyze command to the server.
+ * In async mode, return after sending the command; else, wait for it to finish.
  *
  * Any errors during command execution are reported to stderr.  If async is
  * false, this function exits the program after reporting the error.
  */
 static void
-run_vacuum_command(PGconn *conn, const char *sql, bool echo,
+run_vacuum_command(PGconn *conn, const char* sql, bool echo,
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:1006]: Nit: Unnecessary whitespace diff

File: src/bin/scripts/vacuumdb.c [hunk 12]
```diff
@@ -931,19 +975,35 @@ prepare_vacuum_command(PQExpBuffer sql, int serverVersion,
...
+finalize_vacuum_command(PQExpBuffer sql)
+{
+	appendPQExpBufferChar(sql, ';');
 }
 
 /*
- * Send a vacuum/analyze command to the server.  In async mode, return after
- * sending the command; else, wait for it to finish.
+ * Send a vacuum/analyze command to the server.
+ * In async mode, return after sending the command; else, wait for it to finish.
```
Comment (soumyadeep2007) [inline src/bin/scripts/vacuumdb.c:1000]: Nit: Unnecessary whitespace diff

File: src/bin/scripts/vacuumdb.c (hunk 8 above)
Comment (bmdoil) [reply src/bin/scripts/vacuumdb.c:764]: > After this loop ends you need to process the "last batch" in the case that you had ntables not divisible by jobs.

Right, thanks for catching that. 

> So outside the loop...

Alternatively, we could check if we're at the end of the table list inside the loop. If so, execute the vacuum.
```
                /*
-                * GPDB: When tablesPerConn tables have been processed, execute the VACUUM.
+                * GPDB: When tablesPerConn tables have been processed or we are
+                * at the end of the table list, execute the VACUUM.
                 *
                 * If not in parallel mode, this terminates the
                 * program in case of an error.  (The parallel case handles query
                 * errors in ProcessQueryResult through GetIdleSlot.)
                 */
-               if ((++processedTables == tablesPerConn))
+               if ((++processedTables == tablesPerConn) || (cell->next == NULL))
```
Will execute 4 `VACUUM`

```
statement: VACUUM (SKIP_DATABASE_STATS) public.t7, public.t8;",,,,,,,0,,"postgres.c",1726,
statement: VACUUM (SKIP_DATABASE_STATS) public.t5, public.t6;",,,,,,,0,,"postgres.c",1726,
statement: VACUUM (SKIP_DATABASE_STATS) public.t1, public.t2;",,,,,,,0,,"postgres.c",1726,
statement: VACUUM (SKIP_DATABASE_STATS) public.t3, public.t4;",,,,,,,0,,"postgres.c",1726,
statement: VACUUM (SKIP_DATABASE_STATS) public.t9, public.t10;",,,,,,,0,,"postgres.c",1726,
statement: VACUUM (ONLY_DATABASE_STATS);",,,,,,,0,,"postgres.c",1726,
```

Comment (soumyadeep2007) [reply src/bin/scripts/vacuumdb.c:767]: Sounds great.

**Review by soumyadeep2007 — APPROVED** [review]
LGTM!
